lib/pure/strmisc.nim

```diff
@@ -85,6 +86,110 @@ proc rpartition*(s: string, sep: string): (string, string, string)
+  ##
+  ## The following assumptions and requirements must be met:
+  ## - String must not be empty.
+  ## - String must be stripped of trailing and leading whitespaces.
+  ## - `sep` must not be `'-'`.
+  ## - `decimalDot` must not be `'-'` nor `' '`.
+  ## - `sep` and `decimalDot` must be different.
+  ## - No separator before a digit.
+  ## - First separator can be anywhere after first digit, but no more than 3 characters.
+  ## - There has to be 3 digits between successive separators.
+  ## - There has to be 3 digits between the last separator and the decimal dot.
+  ## - No separator after decimal dot.
+  ## - No duplicate separators.
+  ## - Floats without separator allowed.
+  ##
+  ## See also:
+  ## * `strutils <strutils.html>`_
+  runnableExamples:
+    doAssert parseFloatThousandSep("0") == 0.0
+    doAssert parseFloatThousandSep("-0") == -0.0
+    doAssert parseFloatThousandSep("0.0") == 0.0
+    doAssert parseFloatThousandSep("1.0") == 1.0
+    doAssert parseFloatThousandSep("-0.0") == -0.0
+    doAssert parseFloatThousandSep("-1.0") == -1.0
+    doAssert parseFloatThousandSep("1.000") == 1.0
+    doAssert parseFloatThousandSep("-1.000") == -1.0
+    doAssert parseFloatThousandSep("1,000") == 1000.0
+    doAssert parseFloatThousandSep("-1,000") == -1000.0
+    doAssert parseFloatThousandSep("10,000.000") == 10000.0
+    doAssert parseFloatThousandSep("1,000,000.000") == 1000000.0
+    doAssert parseFloatThousandSep("10,000,000.000") == 10000000.0
+    doAssert parseFloatThousandSep("10.000,0", '.', ',') == 10000.0
+    doAssert parseFloatThousandSep("1'000'000,000", '\'', ',') == 1000000.0
+  assert sep != '-' and decimalDot notin {'-', ' '} and sep != decimalDot
+  if str.len > 1: # Allow "0" thats valid, is 0.0
+    var s = newStringOfCap(str.len)
+    var idx, successive: int
+    var afterDot, lastWasDot, lastWasSep, hasAnySep, isNegative: bool
+    for c in str:
+      if c in '0' .. '9':  # Digits
+        if hasAnySep and successive > 2:
+          raise newException(ValueError,
+            "Invalid float containing thousand separators, more than 3 digits between thousand separators '" &
+            c & "' at index " & $idx)
+        else:
+          s.add c
+          lastWasSep = false
+          lastWasDot = false
+          inc successive
+          inc idx
+      if c == sep:  # Thousands separator, this is NOT the dot
+        if isNegative and idx == 1 or idx == 0:
+          raise newException(ValueError,
+            "Invalid float containing thousand separators, string starts with thousand separator '" &
+            c & "' at index " & $idx)
+        elif lastWasSep:
+          raise newException(ValueError,
+            "Invalid float containing thousand separators, two separators in a row '" &
+            c & "' at index " & $idx)
+        elif afterDot:
+          raise newException(ValueError,
+            "Invalid float containing thousand separators, separator found after decimal dot '" &
```

Comment:
* as @araq would say this causes code bloat; (as a consequence, this can lead to worse performance because of instruction cache)

this can be seen by comparing codegen for:
```nim
when defined case1:
  proc main(a: int)=
    if a == 2: raise newException(ValueError, "ok2")
    elif a == 3: raise newException(ValueError, "ok3")
    elif a == 1: raise newException(ValueError, "ok1")
  main(1)

when defined case2:
  proc raise2(msg: string) =
    raise newException(ValueError, msg)
  proc main(a: int)=
    if a == 2: raise2("ok2")
    elif a == 3: raise2("ok2")
    elif a == 1: raise2("ok1")
  main(1)
```

* you can fix via an auxiliary proc:

```nim
proc raiseError(index: int, input: string) {.noinline, noreturn.} =
  raise newException(ValueError, "Invalid char at position: " & $index & " for input: " & input)
```

note, I'm not sure you really need to specify more detailed error messages (ie 1 generic one like above should suffice), but if you do, then just pass an additional `msg = ""` for more info for cases that need it)